Speed up "Synchronizing DAO" by ~30%

cd2357 opened the pull request:
`BlockParser.parseBlock()` normally takes about 300ms per block.

One part of it is checking if the new block hash is known, which was done by string-comparing it to the hashes of all known blocks. In my case, this was about 110 000 string-comparisons per `BlockParser.parseBlock()` call.

This commit adds a block hash cache in the `DaoState` and replaces the hash string-comparisons by one `HashSet` lookup.

With this change, `BlockParser.parseBlock()` takes around 200ms on my setup.

cd2357 committed:
Improve timer accuracy in BlockParser

Measure the entire parseBlock() duration, not just the substep of parsing transactions.

cd2357 committed:
Cache block hashes in DaoState

Improve isBlockHashKnown() by about 100ms per call by caching the hashes of the known blocks.

cd2357 committed:
Remove duplicate lookup method

Method is duplicate of isBlockHashKnown().

cd2357 committed:
Add blockHeight cache in DaoState

Improve block lookup by height by introducing an index, instead of iterating through all blocks and comparing the height.

cd2357 committed:
Address Codacy complaint / newline

ripcurlx commented:
As by definition this needs an ACK by @ManfredKarrer to get merged.

chimp1984 commented on core/src/main/java/bisq/core/dao/state/model/DaoState.java:
```diff
@@ -109,6 +109,8 @@ public static DaoState getClone(DaoState daoState) {
     @JsonExclude
     private transient final Map<String, Tx> txCache; // key is txId
     @JsonExclude
+    private transient final Map<Integer, Block> blockHeightCache; // key is block height
```
Maybe better name is blocksByHeight.

chimp1984 reviewed (changes requested):
@cd2357 Thanks for working on that.
There is missing the chache update when a new block is added in `onNewBlockWithEmptyTxs`.
Maybe its better to create explicit methods for `daoState.getBlocks().addAll` and `daoState.getBlocks().add` in DaoStateService.

ripcurlx commented:
@chimp1984 Thanks for doing a review as well! But still we'll need in the end the ACK by @ManfredKarrer as well.

chimp1984 commented:
> @chimp1984 Thanks for doing a review as well! But still we'll need in the end the ACK by @ManfredKarrer as well.

Just to be clear, it is a NACK at current state. Cache does not get updated at new blocks.

cd2357 committed:
Strengthen index integrity

Update the two new indices on every known write to the list, namely addBlock() and addBlocks(). Restrict ability to modify the blocks list outside of these methods by overwriting the getBlocks() getter such that it returns an unmodifiable list.

cd2357 commented:
Done, thanks for the speedy review.

> There is missing the chache update when a new block is added in `onNewBlockWithEmptyTxs`.
> Maybe its better to create explicit methods for `daoState.getBlocks().addAll` and `daoState.getBlocks().add` in DaoStateService.

Good point. I restricted the adding of blocks to two new methods, `addBlock` and `addBlocks`. They make sure to update all the indices as well. Also made sure the `DaoState.blocks` list is not exposed directly anymore, but as an unmodifiable list, so no direct modifications are possible.

chimp1984 commented on core/src/main/java/bisq/core/dao/state/model/DaoState.java:
```diff
@@ -109,9 +111,9 @@ public static DaoState getClone(DaoState daoState) {
+    private transient final Map<Integer, Block> blocksByHeight; // Blocks indexed by height
     @JsonExclude
-    private transient final Set<String> blockHashCache;
+    private transient final Set<String> blocksByHash; // Blocks indexed by hash
```
With blocksByHash I would expect a map with hash as key and blocks as value. For a set I would use 'blockHashes'

cd2357 replied:
Agree. I had forgotten one is a set, the other a map.

chimp1984 reviewed (dismissed):
Beside the small naming issue utACK

cd2357 committed:
Rename variable for clarity

Rename blocksByHash because it is not actually indexing blocks by hashes, instead it's a cache of known hashes.

ripcurlx commented:
So only the ACK by @ManfredKarrer is finally missing :smile: :+1:

ManfredKarrer reviewed (approved):
utACK

ripcurlx reviewed (approved):
utACK based on https://github.com/bisq-network/bisq/pull/5484#pullrequestreview-660971809
